### PR TITLE
Version: Add parser for additional hyphenated filenames

### PR DESCRIPTION
**Pull request by samford**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

The version for the `luajit-openresty` formula is currently parsed from the `stable` URL (`https://github.com/openresty/luajit2/archive/refs/tags/v2.1-20210510.tar.gz`) as `20210510` instead of `2.1-20210510` (see Homebrew/homebrew-core#78791). This should have used `version` to specify the full version that wasn't being properly parsed when the formula was added but that wasn't addressed at the time. [Other formulae with a similar version format run into the same parsing issue and they use `version` because of it.]

We could simply add `version "v2.1-20210510"` to `luajit-openresty` (doing what formulae with a similar version format have done) but I figured it would be better to support this version format internally (i.e., fixing this issue for both existing and future formulae without needing to use `version`).

---

This PR addresses this by adding a `StemParser` to `Version::VERSION_PARSERS` that matches versions that start with a numeric version like `1.2`/`v1.2`, `1.2.3`/`v1.2.3`, etc. followed by a hyphen and a trailing number (e.g., `-12`), numeric version (e.g., `-0.2`), date (e.g., `20210605`), etc. This allows these versions to be correctly parsed without needing to use `version` in the formula/cask.

This not only fixes the version parsing for `luajit-openresty` but also for six formulae that are currently using ` version` due to this shortcoming in the current version parsing logic. This would allow us to remove `version` from the following formulae: `bandcamp-dl`, `inxi`, `jnethack`, `luv`, `redir`, `youtube-dlc`.

---

The only other formula or cask that this affects is the `ammonite-repl` formula, which uses a filename like `2.13-2.3.8`, where the numeric version before the hyphen is a Scala version and the numeric version after the hyphen is the `ammonite-repl` version. The aforementioned version parsing bug was unintentionally producing the correct version for this formula because the leading `2.13-` was treated like a software name prefix (e.g., `brew-`) and omitted from the capture group.

Accounting for filenames like `v0.0.8-12.tar.gz` means that we have to use `version` in `ammonite-repl` to manually isolate the formula version. This would need to be done before this PR could be merged.

It's worth mentioning that there are other formulae that have a leading Scala version in the numeric portion of the filename (`akka`, `giter8`, `scalastyle`, `kafka`) but they're unaffected by the change in this PR because they all have a software name prefix before the numeric version (e.g., `akka_2.11-2.4.20.zip`) and the added parser in this PR only matches filenames without one.

### CODE REVIEW

**Comment by carlocab** [comment]
I'm personally not convinced this is needed and forces us to carry around `version_scheme` in `luajit-openresty` when there's zero need for it.

**Review by carlocab — CHANGES_REQUESTED** [review]
I don't think this is needed. Matching the date is enough. The only time it fails is if a project cuts two release tags on the same day, where the second release tag increments the `v\d+.\d+`.

**Comment by samford** [comment]
To me, we either correctly parse the version or we don't. At the moment, we're not correctly parsing the version for `luajit-openresty` and this is only due to a shortcoming in `Version`'s parsing logic (i.e., this isn't an intentional decision).

The trailing date may be sufficient for the foreseeable future (due to upstream's current version constraints) but it's not the full version. Upstream may currently hold the position that they won't release a new major/minor but I don't see any reason to explicitly rely on this, as there's nothing to prevent it from changing in the future. It may be unlikely but it's not impossible.

The only benefit I see in leaving this alone is that we wouldn't have to bump the `version_scheme` for `luajit-openresty`. Maybe I'm being naive but having to bump the `version_scheme` in one formula to fix this doesn't seem like an undue burden.

> It also currently applies to only one formula.

Looking at it from a different perspective, this is one outlier where `brew` isn't reporting the correct, full version and this change would bring it in line with the majority.

**Comment by samford** [comment]
I've updated this to use a more general regex (not specifically tied to a date-based suffix) and this now fixes version parsing for six other formulae as well: `bandcamp-dl`, `inxi`, `jnethack`, `luv`, `redir`, `youtube-dlc`. These currently use `version` in the formula because the previous `Version` parsing behavior didn't produce the correct, full version.

This means that we could remove `version` from these formulae and CI is predictably failing in "tap_syntax (Linux)" with errors like:

```
Error: 6 problems in 6 formulae detected
Error: Stable: version 0.0.8-12 is redundant with version scanned from URL
Error: Stable: version 3.3.04-1 is redundant with version scanned from URL
Error: Stable: version 3.6.6-0.2 is redundant with version scanned from URL
Error: Stable: version 1.41.0-0 is redundant with version scanned from URL
Error: Stable: version 2.2.1-9 is redundant with version scanned from URL
Error: Stable: version 2020.11.11-3 is redundant with version scanned from URL
Error: Process completed with exit code 1.
```

Past that, `ammonite-repl` is parsed differently due to the change in this PR and we would need to use `version` in the formula to ensure the correct version continues to be used. A more lengthy explanation of this situation can be found in the updated PR description but this is expected, as `ammonite-repl` uses an unusual filename format.

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by samford** [comment]
How do we want to handle the `Stable: version X is redundant with version scanned from URL` errors here and in the `ammonite-repl` PR (Homebrew/homebrew-core#78878)?

To maintain the correct version, the `ammonite-repl` PR would need to be merged before this brew PR but it will fail CI until this is merged. Similarly, this brew PR fails CI because it fixes `Version` parsing for six formulae, making their existing `version` statements redundant. Do we simply accept that CI will fail or is there another way around this?

The `luajit-openresty` PR that bumps the `version_scheme` (Homebrew/homebrew-core#78955) will need to be merged after this brew PR. I'll also create a follow-up homebrew/core PR to remove the redundant `version` statements afterward. Am I correct in assuming we'll need to wait to merge these particular PRs until this change is in a tagged `brew` release?

**Comment by MikeMcQuaid** [comment]
> To maintain the correct version, the `ammonite-repl` PR would need to be merged before this brew PR but it will fail CI until then. Similarly, this brew PR fails CI because it fixes `Version` parsing for six formulae, making their existing `version` statements redundant. Do we simply accept that CI will fail or is there a way to get these PRs to pass?

Yes, there's a bit of a chicken-and-egg here that I'm not sure how best to work around. My suggestion would be to amend `brew audit` so that the `is redundant with version scanned from URL` is not output when running `brew audit` on an entire tap but only on a single formula. This would allow this PR to be merged. Thoughts?

**Comment by samford** [comment]
> My suggestion would be to amend `brew audit` so that the `is redundant with version scanned from URL` is not output when running `brew audit` on an entire tap but only on a single formula. This would allow this PR to be merged.

I tinkered with this and the way I approached it was to modify `ResourceAuditor` to allow for `only` and `except` options, borrowing related code from `FormulaAuditor`. This allows the `brew audit` `--only` and `--except` options to apply to `ResourceAuditor`, not just `FormulaAuditor`.

[One thing to note is that an audit with the same name in `FormulaAuditor` and `ResourceAuditor` would both be affected by `--only` or `--except`. There aren't any overlapping names right now but, depending on the audit, it could be either desirable or undesirable behavior.]

```patch
diff --git a/Library/Homebrew/formula_auditor.rb b/Library/Homebrew/formula_auditor.rb
index 06b54db3c0a1ac8dd5b34f6ba39271f2e42e7da6..19be6c03b58503e61f56b061e074e691b9a26c08 100644
--- a/Library/Homebrew/formula_auditor.rb
+++ b/Library/Homebrew/formula_auditor.rb
@@ -500,7 +500,7 @@ module Homebrew
         spec_name = name.downcase.to_sym
         next unless (spec = formula.send(spec_name))
 
-        ra = ResourceAuditor.new(spec, spec_name, online: @online, strict: @strict).audit
+        ra = ResourceAuditor.new(spec, spec_name, online: @online, strict: @strict, only: @only, except: @except).audit
         ra.problems.each do |message|
           problem "#{name}: #{message}"
         end
@@ -508,7 +508,7 @@ module Homebrew
         spec.resources.each_value do |resource|
           problem "Resource name should be different from the formula name" if resource.name == formula.name
 
-          ra = ResourceAuditor.new(resource, spec_name, online: @online, strict: @strict).audit
+          ra = ResourceAuditor.new(resource, spec_name, online: @online, strict: @strict, only: @only, except: @except).audit
           ra.problems.each do |message|
             problem "#{name} resource #{resource.name.inspect}: #{message}"
           end
diff --git a/Library/Homebrew/resource_auditor.rb b/Library/Homebrew/resource_auditor.rb
index f18652715fcfc4ddda5958271e49990a23742f60..bfc71517ca06d6cd0f3edde5a63d610d3192455e 100644
--- a/Library/Homebrew/resource_auditor.rb
+++ b/Library/Homebrew/resource_auditor.rb
@@ -20,14 +20,25 @@ module Homebrew
       @spec_name = spec_name
       @online    = options[:online]
       @strict    = options[:strict]
+      @only      = options[:only]
+      @except    = options[:except]
       @problems  = []
     end
 
     def audit
-      audit_version
-      audit_download_strategy
-      audit_checksum
-      audit_urls
+      only_audits = @only
+      except_audits = @except
+
+      methods.map(&:to_s).grep(/^audit_/).each do |audit_method_name|
+        name = audit_method_name.delete_prefix("audit_")
+        if only_audits
+          next unless only_audits.include?(name)
+        elsif except_audits
+          next if except_audits.include?(name)
+        end
+        send(audit_method_name)
+      end
+
       self
     end
 
```

---

With these changes in place, we could handle this issue in one of two ways:

A) Modify `dev-cmd/audit.rb` to automatically exclude `ResourceAuditor#audit_version` when `brew audit --tap` is used:

```patch
diff --git a/Library/Homebrew/dev-cmd/audit.rb b/Library/Homebrew/dev-cmd/audit.rb
index b7e1bfb433ba2a80e597dbe6cb33dfdf83ea8c90..b97aaf0fe80b0e7df099ba365027ef8d2fee1250 100644
--- a/Library/Homebrew/dev-cmd/audit.rb
+++ b/Library/Homebrew/dev-cmd/audit.rb
@@ -182,7 +182,16 @@ module Homebrew
         display_cop_names:    args.display_cop_names?,
       }.compact
 
-      fa = FormulaAuditor.new(f, **options)
+      fa = if args.tap
+        # Skip ResourceAuditor#audit_version when auditing a tap
+        FormulaAuditor.new(
+          f,
+          except: args.except&.union(["version"]) || ["version"],
+          **options
+        )
+      else
+        FormulaAuditor.new(f, **options)
+      end
       fa.audit
 
       if fa.problems.any? || fa.new_formula_problems.any?

```

The downside is that this applies wherever `brew audit --tap` is used, not just on CI. I feel like maintainers/contributors should be able to surface redundant versions when running `brew audit --tap` locally and the exclusion should only happen on CI/test-bot. That is to say, we shouldn't have to run `brew audit` on specific formulae to get the version audit.

B) Modify related GitHub Actions workflows and test-bot to use the `--except=version` option for `brew audit` steps (e.g., the "Run brew audit --skip-style on all taps" step of the Homebrew/brew `tests.yml` workflow). This only impacts CI/test-bot and doesn't affect local `brew audit --tap=` behavior, so this may be a better solution if we want to keep this around long term.

Homebrew/brew:

```patch
diff --git a/.github/workflows/tests.yml b/.github/workflows/tests.yml
index 12c6f8444b8f37914a263f04b9d77b4a356f8547..e77235d71046829640cdb2532d1a4d7322581a23 100644
--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -78,7 +78,7 @@ jobs:
         run: brew style --display-cop-names homebrew/core
 
       - name: Run brew audit --skip-style on all taps
-        run: brew audit --skip-style --display-failures-only
+        run: brew audit --display-failures-only --skip-style --except=version
 
       - name: Set up all Homebrew taps
         run: |

```

test-bot:

```patch
diff --git a/lib/tests/tap_syntax.rb b/lib/tests/tap_syntax.rb
index 661638e6243df2a5b6f34fbd7ddfa180c7058a7e..482076a7e427a6cf62dec4904769f78717528b86 100644
--- a/lib/tests/tap_syntax.rb
+++ b/lib/tests/tap_syntax.rb
@@ -12,7 +12,7 @@ module Homebrew
                                 MacOS.active_developer_dir == "/Applications/Xcode.app/Contents/Developer"
         test "brew", "style", tap.name unless broken_xcode_rubygems
 
-        test "brew", "audit", "--tap=#{tap.name}"
+        test "brew", "audit", "--tap=#{tap.name}", "--except=version"
       end
     end
   end

```

---

What do you think about these approaches?

Depending on how we handle this, these audit changes don't necessarily have to stick around long-term. However, if we revert the changes after merging related PRs, we would have to do this dance again the next time we modify `Version` parsing.

That said, I'll create a PR once we decide on how to approach this and we can go from there.

**Comment by MikeMcQuaid** [comment]
> B) Modify related GitHub Actions workflows and test-bot to use the `--except=version` option for `brew audit` steps (e.g., the "Run brew audit --skip-style on all taps" step of the Homebrew/brew `tests.yml` workflow). This only impacts CI/test-bot and doesn't affect local `brew audit --tap=` behavior, so this may be a better solution if we want to keep this around long term.

This feels like the best approach to me 👍🏻 

> That said, I'll create a PR once we decide on how to approach this and we can go from there.

Perfect, thanks @samford. I love how well you spelled out both of these options. You're making my life easy here 😆🏝

**Comment by samford** [comment]
> You're making my life easy here 😆🏝

I feel like I make enough work for you in PR review, so I try to make it easy when I can, ahah.

---

brew and test-bot PRs are now up: #11542, Homebrew/homebrew-test-bot#617.

**Comment by MikeMcQuaid** [comment]
> I feel like I make enough work for you in PR review, so I try to make it easy when I can, ahah.

Haha, I feel more guilty about my PR reviews. Thanks again @samford ❤️

**Comment by samford** [comment]
With #11542 merged, this now passes CI but we can't merge it until the `ammonite-repl` PR (Homebrew/homebrew-core#78878) is merged. We can't merge that until the `CI-syntax-only` label is working again in homebrew/core (it broke with the changes made to `.github/workflows/test.yml` earlier today), as a full CI run will fail without the [additional test-bot change](https://github.com/Homebrew/homebrew-test-bot/pull/617#discussion_r652546506) I omitted in response to feedback.

So close!

**Comment by samford** [comment]
The `CI-syntax-only` fix was merged and the homebrew/core PR to preemptively add a `version` to `ammonite-repl` passed CI after being rebased, so that's now merged. I'm going to go ahead and merge this PR now, so it will be included in the next tagged `brew` version.

Once it's in a tagged version, I'll merge the `luajit-openresty` PR (Homebrew/homebrew-core#78955) and create a PR to remove the newly-redundant `version` calls from the aforementioned formulae.

**Pull request merged**
